Support title attribute for pattern validation errors

annevk opened the pull request:
#### 2d36f436ba4efb45b5e3f3826637e671c9b13eaa
<pre>
Support title attribute for pattern validation errors
<a href="https://bugs.webkit.org/show_bug.cgi?id=258301">https://bugs.webkit.org/show_bug.cgi?id=258301</a>
rdar://111394402

Reviewed by Aditya Keerthi.

The HTML Standard suggests that the title attribute can be used to
augment the message presented to the end user. As this is not trusted
UI space this seems reasonable (JavaScript has full control over this
UI) and also matches other browsers.

* LayoutTests/fast/forms/validation-messages-expected.txt:
* LayoutTests/fast/forms/validation-messages.html:
* LayoutTests/platform/ios-wk2/fast/forms/validation-messages-expected.txt:
* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/html/InputType.cpp:
(WebCore::InputType::validationMessage const):
* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::validationMessagePatternMismatchText):
* Source/WebCore/platform/LocalizedStrings.h:

Canonical link: <a href="https://commits.webkit.org/266311@main">https://commits.webkit.org/266311@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dfdf34ef5cab4bddbeaf858bc1dfc0939eeb9ba1

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/13514 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/13828 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/14159 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/15251 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/12854 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/13593 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/16336 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/13905 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/15539 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/13681 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/14328 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/11440 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/15955 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/11615 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/12195 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/19242 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/12691 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/12363 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/15577 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/12875 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/10759 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/12143 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/12060 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3285 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/16471 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/12718 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aestes commented on Source/WebCore/en.lproj/Localizable.strings:
```diff
@@ -778,6 +778,12 @@
+"Match the requested format: %@" = "Match the requested format: %@";
+
+/* Validation message for input form controls requiring a constrained value according to pattern followed by a website-provided description of the pattern */
+"Match the requested format: %s" = "Match the requested format: %s";
```
Is the `%s` variant really needed here? I think this strings file is only used by the Cocoa platforms.

annevk replied:
While this was copypasta, the fact that `update-webkit-localizable-strings` ends up doing the exact same makes me inclined to leave this as-is.

pxlcoder replied:
@WebKit/glib-reviewers Is this string needed for localization on your platforms?

mcatanzaro replied:
Yes: it's displayed to users so it needs to be localized. Marking it with WEB_UI_FORMAT_STRING is all you need to do to achieve that. i.e. it's already done properly.

This Localizable.strings is used only by Apple and Windows platforms, but you might as well leave it since it will just be added again next time the script is run.

pxlcoder replied:
Thanks for confirming! We should probably fix the script at some point...

annevk replied:
Filed https://bugs.webkit.org/show_bug.cgi?id=259519.

pxlcoder commented on Source/WebCore/html/InputType.cpp:
```diff
@@ -564,8 +564,12 @@ String InputType::validationMessage() const
-    if (patternMismatch(value))
-        return validationMessagePatternMismatchText();
+    if (patternMismatch(value)) {
+        auto title = element()->getAttribute(HTMLNames::titleAttr).string().trim(isASCIIWhitespace).simplifyWhiteSpace(isASCIIWhitespace);
```
```suggestion
        auto title = element()->attributeWithoutSynchronization(HTMLNames::titleAttr).string().trim(isASCIIWhitespace).simplifyWhiteSpace(isASCIIWhitespace);
```

annevk committed:
Support title attribute for pattern validation errors
https://bugs.webkit.org/show_bug.cgi?id=258301
rdar://111394402

Reviewed by Aditya Keerthi.

The HTML Standard suggests that the title attribute can be used to
augment the message presented to the end user. As this is not trusted
UI space this seems reasonable (JavaScript has full control over this
UI) and also matches other browsers.

* LayoutTests/fast/forms/validation-messages-expected.txt:
* LayoutTests/fast/forms/validation-messages.html:
* LayoutTests/platform/ios-wk2/fast/forms/validation-messages-expected.txt:
* Source/WebCore/en.lproj/Localizable.strings:
* Source/WebCore/html/InputType.cpp:
(WebCore::InputType::validationMessage const):
* Source/WebCore/platform/LocalizedStrings.cpp:
(WebCore::validationMessagePatternMismatchText):
* Source/WebCore/platform/LocalizedStrings.h:

Canonical link: https://commits.webkit.org/266311@main
